printer/printer.go

```diff
@@ -0,0 +1,188 @@
+package printer
+
+import (
+	"fmt"
+
+	ast_pb "github.com/unpackdev/protos/dist/go/ast"
+	"github.com/unpackdev/solgo/ast"
+)
+
+// Printer prints out the code based on the AST
+type Printer struct {
+	output string
+}
+
+func New() *Printer {
+	return &Printer{
+		output: "",
+	}
+}
+
+func (p *Printer) Output() string {
+	return p.output
+}
+
+// Print traverses the AST and prints the nodes
+func (p *Printer) Print(root *ast.RootNode) {
+	nodes := root.GetNodes()
+	p.traverseNodes(nodes, 0)
+}
+
+func (p *Printer) traverseNodes(nodes []ast.Node[ast.NodeType], depth int) {
+	for _, node := range nodes {
+		p.visitNode(node)
+		childNodes := node.GetNodes()
+		if len(childNodes) > 0 {
+			p.traverseNodes(childNodes, depth+1)
+		}
+	}
+}
+
+func (p *Printer) visitNode(node ast.Node[ast.NodeType]) {
```

Comment:
Typecasting is necessary in the `visitNode` method because the `node` parameter is of type `ast.Node[ast.NodeType]`, which is a generic interface type. The interface type doesn't provide access to the specific fields and methods of the concrete node types.

To access the specific fields and methods of each node type, you need to typecast the `node` to its corresponding concrete type. This is done using the type assertion syntax `node.(*ConcreteType)`.

For example, when the `node.GetType()` returns `ast_pb.NodeType_SOURCE_UNIT`, the code typecasts `node` to `*ast.SourceUnit[ast.Node[ast_pb.SourceUnit]]`:

```go
n := node.(*ast.SourceUnit[ast.Node[ast_pb.SourceUnit]])
```

By typecasting `node` to `*ast.SourceUnit[ast.Node[ast_pb.SourceUnit]]`, you can access the specific fields and methods of the `SourceUnit` struct, such as `n.Name` or `n.GetNodes()`.

Similarly, for other node types like `ast_pb.NodeType_PRAGMA_DIRECTIVE`, `ast_pb.NodeType_IMPORT_DIRECTIVE`, etc., the code typecasts `node` to their corresponding concrete types (`*ast.Pragma`, `*ast.Import`, etc.) to access their specific fields and methods.

Typecasting is necessary because the `visitNode` method receives a generic `ast.Node[ast.NodeType]` interface type, and to perform specific operations on each node type, you need to convert it to its concrete type.

Without typecasting, you would only have access to the methods defined in the `ast.Node[ast.NodeType]` interface, which may not provide the necessary functionality for printing or processing each node type specifically.

So, typecasting allows you to treat the generic `ast.Node[ast.NodeType]` as its specific concrete type, enabling you to access and utilize the fields and methods specific to each node type in the respective visitor methods.

Reply:
Guys, jumping in to mention that you can now use https://github.com/unpackdev/solgo/blob/4ffb355a8d82e06fae608b1540755717c30c40c2/ir/root.go#L234

Basically you can walk recursively through the entire tree however you wish. There's new NodeVisitor -> https://github.com/unpackdev/solgo/blob/main/ast/visitor.go

Reply:
It's released under latest Brejcha 0.3.3 including many other things if you need it.